File: 백준/Gold/13023. ABCDE/ABCDE.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    static ArrayList<ArrayList<Integer>> graph;
    static boolean isConnected;
    static boolean[] visited;

    static void dfs(int node, int depth) {
        if (depth == 5 || isConnected) {
            isConnected = true;
            return;
        }

        visited[node] = true;

        for (int v : graph.get(node)) {
            if (!visited[v]) dfs(v, depth + 1);
        }

        visited[node] = false;
    }

    static void addEdge(int u, int v) {
        graph.get(u).add(v);
        graph.get(v).add(u);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        visited = new boolean[n];
        graph = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            addEdge(u, v);
        }

        for (int i = 0; i < n; i++) {
            dfs(i, 1);
            if (isConnected) break;
        }

        if (isConnected) System.out.println(1);
        else System.out.println(0);
    }
}

File: 백준/Gold/13023. ABCDE/ABCDE.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer

var isConnected = false
lateinit var graph: MutableList<MutableList<Int>>
lateinit var visited: BooleanArray

fun dfs(node: Int, depth: Int) {
    if (depth == 5 || isConnected) {
        isConnected = true
        return
    }

    visited[node] = true

    for (v in graph[node]) {
        if (!visited[v]) dfs(v, depth + 1)
    }

    visited[node] = false
}

fun addEdge(u: Int, v: Int) {
    graph[u].add(v)
    graph[v].add(u)
}

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine())
    val n = st.nextToken().toInt()
    val m = st.nextToken().toInt()
    visited = BooleanArray(n)
    graph = MutableList(n) { mutableListOf() }

    repeat(m) {
        st = StringTokenizer(br.readLine())
        val u = st.nextToken().toInt()
        val v = st.nextToken().toInt()
        addEdge(u, v)
    }

    for (i in 0 until n) {
        dfs(i, 1)
        if (isConnected) break
    }

    println(if (isConnected) 1 else 0)
}